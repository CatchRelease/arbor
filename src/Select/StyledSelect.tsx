import { FC } from 'react';
import styled from '@emotion/styled';

import StyledInput from '../Input/StyledInput';

const StyledSelectInput = styled(StyledInput)`
  appearance: none;
  background: ${(props) => props.theme.colors.background.white};

  &:-moz-focusring {
    color: transparent;
    text-shadow: 0 0 0 #000;
  }
`;

StyledSelectInput.defaultProps = {
  as: 'select'
};

const StyledSelectWrapper = styled.div`
  position: relative;

  &:after {
    align-items: center;
    bottom: 0;
    color: ${(props) => props.theme.colors.icon.default};
    content: '${(props) => props.theme.icons.caret}';
    display: flex;
    font-family: ${(props) => props.theme.fonts.icon};
    font-size: ${(props) => props.theme.fontSizes.size5};
    pointer-events: none;
    position: absolute;
    right: ${(props) => props.theme.space.smaller};
    top: 0;
  }
`;

const StyledSelect: FC = ({ children, ...props }) => {
  return (
    <StyledSelectWrapper>
      <StyledSelectInput {...props}>{children}</StyledSelectInput>
    </StyledSelectWrapper>
  );
};

export default StyledSelect;
