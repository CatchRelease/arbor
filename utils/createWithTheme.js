import React from 'react';
import { CacheProvider } from '@emotion/core';
import createCache from '@emotion/cache';
import { create } from 'react-test-renderer';
import { ThemeProvider } from 'emotion-theming';
import theme from '../src/theme';

const myCache = createCache();
myCache.compat = true;

export default (Component) =>
  create(
    <CacheProvider value={myCache}>
      <ThemeProvider theme={theme}>{Component}</ThemeProvider>
    </CacheProvider>
  );
