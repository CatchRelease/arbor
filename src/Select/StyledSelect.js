import React from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';

import { createWithComponent } from '../utils';
import StyledInput from '../Input/StyledInput';

const StyledSelectInput = styled(createWithComponent(StyledInput, 'select'))`
  appearance: none;
`;

const StyledSelectWrapper = styled.div`
  position: relative;

  &:after {
    align-items: center;
    bottom: 0;
    color: ${props => props.theme.colors.icon.default};
    content: '${props => props.theme.icons.caret}';
    display: flex;
    font-family: ${props => props.theme.fonts.icon};
    font-size: ${props => props.theme.fontSizes.size5};
    position: absolute;
    right: ${props => props.theme.space.smaller};
    top: 0;
  }
`;

const StyledSelect = ({ children, ...props }) => {
  const { large } = props;

  return (
    <StyledSelectWrapper large={large}>
      <StyledSelectInput {...props}>{children}</StyledSelectInput>
    </StyledSelectWrapper>
  );
};

StyledSelect.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.node,
    PropTypes.arrayOf(PropTypes.node)
  ]).isRequired
};

export default StyledSelect;
