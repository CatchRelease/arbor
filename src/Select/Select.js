import React from 'react';
import PropTypes from 'prop-types';

import FormField from '../FormField';
import StyledSelect from './StyledSelect';

const Select = React.forwardRef(
  ({ caption, children, label, labelAside, id, ...props }, ref) => {
    const select = (
      <StyledSelect {...{ ...props, id, ref }}>{children}</StyledSelect>
    );

    return caption || label ? (
      <FormField
        caption={caption}
        id={id}
        label={label}
        labelAside={labelAside}
      >
        {select}
      </FormField>
    ) : (
      select
    );
  }
);

Select.propTypes = {
  /**
   * Help text to be displayed in the FormField for the Select.
   */
  caption: PropTypes.string,

  /**
   * Select options to be displayed within the menu
   */
  children: PropTypes.oneOfType([
    PropTypes.node,
    PropTypes.arrayOf(PropTypes.node)
  ]).isRequired,

  /**
   * HTML ID for the select
   */
  id: PropTypes.string,

  /**
   * Label text for the select
   */
  label: PropTypes.oneOfType([PropTypes.node, PropTypes.string]),

  /**
   * Component that goes to the right of the label. Does not have to only be text.
   */
  labelAside: PropTypes.node
};

Select.defaultProps = {
  id: undefined,
  caption: '',
  label: '',
  labelAside: null
};

export default Select;
