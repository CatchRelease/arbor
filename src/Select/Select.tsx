import { FC, forwardRef, ReactNode } from 'react';

import FormField from '../FormField';
import StyledSelect from './StyledSelect';

type Props = {
  caption?: string;
  id: string;
  label?: ReactNode;
  labelAside?: ReactNode;
};

const Select: FC<Props> = forwardRef<HTMLSelectElement, Props>(
  (
    { caption = '', children, label = '', labelAside = null, id, ...props },
    ref
  ) => {
    const select = (
      <StyledSelect {...{ ...props, id, ref }}>{children}</StyledSelect>
    );

    return caption || label ? (
      <FormField
        caption={caption}
        id={id}
        label={label}
        labelAside={labelAside}
      >
        {select}
      </FormField>
    ) : (
      select
    );
  }
);

export default Select;
