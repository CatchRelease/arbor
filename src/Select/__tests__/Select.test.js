import React from 'react';

import createWithTheme from '../../../utils/createWithTheme';
import Select from '../Select';

const render = (props = {}) =>
  createWithTheme(
    <Select id="example" {...props}>
      <option>Some Option</option>
    </Select>
  ).toJSON();

describe('<Select />', () => {
  it('renders an Select correctly', () => {
    const tree = render();

    expect(tree).toMatchSnapshot();
  });

  it('renders a large Select correctly', () => {
    const tree = render({ large: true });

    expect(tree).toMatchSnapshot();
  });

  it('renders a Select with a caption correctly', () => {
    const tree = render({
      caption: 'I am a caption'
    });

    expect(tree).toMatchSnapshot();
  });

  it('renders a Select with a label correctly', () => {
    const tree = render({
      label: 'My Select'
    });

    expect(tree).toMatchSnapshot();
  });

  it('renders a Select with a labelAside correctly', () => {
    const labelAside = <div>hello</div>;

    const tree = render({
      labelAside
    });

    expect(tree).toMatchSnapshot();
  });
});
