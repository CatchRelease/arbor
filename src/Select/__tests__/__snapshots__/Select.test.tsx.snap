// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> renders a Select with a caption correctly 1`] = `
.emotion-0 {
  margin-bottom: 0.5rem;
}

.emotion-1 {
  font-family: Motiva Sans;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-2 {
  position: relative;
}

.emotion-2:after {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  bottom: 0;
  color: #989A9C;
  content: '037';
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-family: Arbor;
  font-size: 1rem;
  pointer-events: none;
  position: absolute;
  right: 0.5rem;
  top: 0;
}

.emotion-3 {
  border-radius: 3px;
  border: 1px solid #E0E1E2;
  box-sizing: border-box;
  color: #25272D;
  font-family: Motiva Sans;
  font-size: 0.875rem;
  line-height: 1.25;
  margin: 0.25rem 0;
  padding: 0.5rem;
  width: 100%;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  background: #FFFFFF;
}

@media (max-width: 768px) {
  .emotion-3 {
    font-size: 1rem;
  }
}

.emotion-3::-webkit-input-placeholder {
  color: #989A9C;
}

.emotion-3::-moz-placeholder {
  color: #989A9C;
}

.emotion-3:-ms-input-placeholder {
  color: #989A9C;
}

.emotion-3::placeholder {
  color: #989A9C;
}

.emotion-3:focus {
  border: 1px solid #081D17;
  box-shadow: 0 0 8px rgba(182,239,208,0.1);
  outline: none;
}

.emotion-3:disabled {
  background: #F2F2F2;
}

.emotion-3:required {
  box-shadow: none;
}

.emotion-3:-moz-focusring {
  color: transparent;
  text-shadow: 0 0 0 #000;
}

.emotion-4 {
  font-family: Motiva Sans;
  color: #D93838;
  line-height: 1.5;
  font-size: 0.8125rem;
}

<div
  className="emotion-0"
>
  <div
    className="emotion-1"
  >
    
  </div>
  <div
    className="emotion-2"
  >
    <select
      className="emotion-3"
      id="example"
    >
      <option>
        Some Option
      </option>
    </select>
  </div>
  <span
    className="emotion-4"
    color="palette.red.default"
  >
    I am a caption
  </span>
</div>
`;

exports[`<Select /> renders a Select with a label correctly 1`] = `
.emotion-0 {
  margin-bottom: 1rem;
}

.emotion-1 {
  font-family: Motiva Sans;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-2 {
  font-family: Motiva Sans;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.emotion-3 {
  color: #5B5D61;
  display: block;
  font-family: Motiva Sans;
  font-size: 0.875rem;
  line-height: 1.5;
}

.emotion-4 {
  position: relative;
}

.emotion-4:after {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  bottom: 0;
  color: #989A9C;
  content: '037';
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-family: Arbor;
  font-size: 1rem;
  pointer-events: none;
  position: absolute;
  right: 0.5rem;
  top: 0;
}

.emotion-5 {
  border-radius: 3px;
  border: 1px solid #E0E1E2;
  box-sizing: border-box;
  color: #25272D;
  font-family: Motiva Sans;
  font-size: 0.875rem;
  line-height: 1.25;
  margin: 0.25rem 0;
  padding: 0.5rem;
  width: 100%;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  background: #FFFFFF;
}

@media (max-width: 768px) {
  .emotion-5 {
    font-size: 1rem;
  }
}

.emotion-5::-webkit-input-placeholder {
  color: #989A9C;
}

.emotion-5::-moz-placeholder {
  color: #989A9C;
}

.emotion-5:-ms-input-placeholder {
  color: #989A9C;
}

.emotion-5::placeholder {
  color: #989A9C;
}

.emotion-5:focus {
  border: 1px solid #081D17;
  box-shadow: 0 0 8px rgba(182,239,208,0.1);
  outline: none;
}

.emotion-5:disabled {
  background: #F2F2F2;
}

.emotion-5:required {
  box-shadow: none;
}

.emotion-5:-moz-focusring {
  color: transparent;
  text-shadow: 0 0 0 #000;
}

<div
  className="emotion-0"
>
  <div
    className="emotion-1"
  >
    <div
      className="emotion-2"
    >
      <label
        className="emotion-3"
        htmlFor="example"
      >
        My Select
      </label>
    </div>
  </div>
  <div
    className="emotion-4"
  >
    <select
      className="emotion-5"
      id="example"
    >
      <option>
        Some Option
      </option>
    </select>
  </div>
  
</div>
`;

exports[`<Select /> renders a Select with a labelAside correctly 1`] = `
.emotion-0 {
  position: relative;
}

.emotion-0:after {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  bottom: 0;
  color: #989A9C;
  content: '037';
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-family: Arbor;
  font-size: 1rem;
  pointer-events: none;
  position: absolute;
  right: 0.5rem;
  top: 0;
}

.emotion-1 {
  border-radius: 3px;
  border: 1px solid #E0E1E2;
  box-sizing: border-box;
  color: #25272D;
  font-family: Motiva Sans;
  font-size: 0.875rem;
  line-height: 1.25;
  margin: 0.25rem 0;
  padding: 0.5rem;
  width: 100%;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  background: #FFFFFF;
}

@media (max-width: 768px) {
  .emotion-1 {
    font-size: 1rem;
  }
}

.emotion-1::-webkit-input-placeholder {
  color: #989A9C;
}

.emotion-1::-moz-placeholder {
  color: #989A9C;
}

.emotion-1:-ms-input-placeholder {
  color: #989A9C;
}

.emotion-1::placeholder {
  color: #989A9C;
}

.emotion-1:focus {
  border: 1px solid #081D17;
  box-shadow: 0 0 8px rgba(182,239,208,0.1);
  outline: none;
}

.emotion-1:disabled {
  background: #F2F2F2;
}

.emotion-1:required {
  box-shadow: none;
}

.emotion-1:-moz-focusring {
  color: transparent;
  text-shadow: 0 0 0 #000;
}

<div
  className="emotion-0"
>
  <select
    className="emotion-1"
    id="example"
  >
    <option>
      Some Option
    </option>
  </select>
</div>
`;

exports[`<Select /> renders a large Select correctly 1`] = `
.emotion-0 {
  position: relative;
}

.emotion-0:after {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  bottom: 0;
  color: #989A9C;
  content: '037';
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-family: Arbor;
  font-size: 1rem;
  pointer-events: none;
  position: absolute;
  right: 0.5rem;
  top: 0;
}

.emotion-1 {
  border-radius: 3px;
  border: 1px solid #E0E1E2;
  box-sizing: border-box;
  color: #25272D;
  font-family: Motiva Sans;
  font-size: 0.875rem;
  line-height: 1.25;
  margin: 0.25rem 0;
  padding: 0.5rem;
  width: 100%;
  padding: 0.75rem 0.5rem;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  background: #FFFFFF;
}

@media (max-width: 768px) {
  .emotion-1 {
    font-size: 1rem;
  }
}

.emotion-1::-webkit-input-placeholder {
  color: #989A9C;
}

.emotion-1::-moz-placeholder {
  color: #989A9C;
}

.emotion-1:-ms-input-placeholder {
  color: #989A9C;
}

.emotion-1::placeholder {
  color: #989A9C;
}

.emotion-1:focus {
  border: 1px solid #081D17;
  box-shadow: 0 0 8px rgba(182,239,208,0.1);
  outline: none;
}

.emotion-1:disabled {
  background: #F2F2F2;
}

.emotion-1:required {
  box-shadow: none;
}

.emotion-1:-moz-focusring {
  color: transparent;
  text-shadow: 0 0 0 #000;
}

<div
  className="emotion-0"
>
  <select
    className="emotion-1"
    id="example"
  >
    <option>
      Some Option
    </option>
  </select>
</div>
`;

exports[`<Select /> renders an Select correctly 1`] = `
.emotion-0 {
  position: relative;
}

.emotion-0:after {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  bottom: 0;
  color: #989A9C;
  content: '037';
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-family: Arbor;
  font-size: 1rem;
  pointer-events: none;
  position: absolute;
  right: 0.5rem;
  top: 0;
}

.emotion-1 {
  border-radius: 3px;
  border: 1px solid #E0E1E2;
  box-sizing: border-box;
  color: #25272D;
  font-family: Motiva Sans;
  font-size: 0.875rem;
  line-height: 1.25;
  margin: 0.25rem 0;
  padding: 0.5rem;
  width: 100%;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  background: #FFFFFF;
}

@media (max-width: 768px) {
  .emotion-1 {
    font-size: 1rem;
  }
}

.emotion-1::-webkit-input-placeholder {
  color: #989A9C;
}

.emotion-1::-moz-placeholder {
  color: #989A9C;
}

.emotion-1:-ms-input-placeholder {
  color: #989A9C;
}

.emotion-1::placeholder {
  color: #989A9C;
}

.emotion-1:focus {
  border: 1px solid #081D17;
  box-shadow: 0 0 8px rgba(182,239,208,0.1);
  outline: none;
}

.emotion-1:disabled {
  background: #F2F2F2;
}

.emotion-1:required {
  box-shadow: none;
}

.emotion-1:-moz-focusring {
  color: transparent;
  text-shadow: 0 0 0 #000;
}

<div
  className="emotion-0"
>
  <select
    className="emotion-1"
    id="example"
  >
    <option>
      Some Option
    </option>
  </select>
</div>
`;
